bionic/code_references.py

```diff
@@ -0,0 +1,214 @@
+
+
+@attr.s
+class CodeContext:
+    """
+    Holds variable information from function and code object
+    attributes.
+
+    Attributes
+    ----------
+    globals: dict
+        A reference to the dictionary that holds the function’s global
+        variables — the global namespace of the module in which the
+        function was defined.
+    cells: dict
+        A dictionary that tracks all free variables by name and cell
+        variable names by themselves.
+    varnames: dict
+        A dictionary that is used to track local variables by name.
+    """
+
+    globals = attr.ib()
+    cells = attr.ib()
+    varnames = attr.ib()
+
+
+def get_code_context(func) -> CodeContext:
+    code = func.__code__
+
+    # Mapping from variable name to the value if we can resolve it.
+    # Otherwise map to the name.
+    cells = {}
+
+    for var in code.co_cellvars:
+        # Streamlit uses the names of the variable instead of the value
+        # for cell variables. I suspect this is so because it is hard
+        # to find the value of a cell variable. It doesn't mean we
+        # don't hash these values for Smart Caching. These cell
+        # variables are also present in ``code.co_consts``` which we
+        # hash separately.
+        cells[var] = var
+
+    if code.co_freevars:
+        assert len(code.co_freevars) == len(func.__closure__)
+        for freevar, cell in zip(code.co_freevars, func.__closure__):
+            cells[freevar] = cell.cell_contents
+
+    varnames = {}
+    if inspect.ismethod(func):
+        varnames = {"self": func.__self__}
+
+    return CodeContext(globals=func.__globals__, cells=cells, varnames=varnames)
+
+
+def get_referenced_objects(code, context):
+    """
+    Returns referenced objects for a code object and name of the references when
+    the references cannot be detected without performing runtime logic.
+
+    Referenced objects can be anything from a class, a function, a module or any
+    variables. The references can also be any scope, like global, local, cell,
+    free, or can even be in another object referenced using a full qualified
+    name.
+
+    Note that this function does not perform any runtime logic. This includes
+    calling any functions used by the input code because doing so can be
+    expensive and have unintended consequences. Due to this, any references with
+    a fully qualified name that uses result of such a call won't be detected.
+    This means that if an inner function returns a module or a class, any
+    variables of the module or the class won't be detected. In this case, we will
+    return the name of the reference as a proxy. So for a function like below,
+    the returned reference would be ``"call"`` instead of ``MyClass.call``.
+
+    .. code-block:: python
+
+        def x():
+            my_cls = get_my_class("MyClass") # Returns class MyClass
+            my_cls.call()
+
+
+    On a higher level, for Smart Caching, this leaves some changes that Bionic
```

Comment:
I don't think "Smart Caching" is referenced anywhere else in our codebase. I don't think you need to go into a lot of detail here, but can we write this in a more standalone way?